dask/highlevelgraph.py

```diff
@@ -24,6 +27,109 @@ def _find_layer_containing_key(key):
     return ret
 
 
+class Layer(collections.abc.Mapping):
+    """High level graph layer
+
+    This abstract class establish a protocol for high level graph layers.
+    """
+
+    def cull(
+        self, keys: Set, all_hlg_keys: Container
+    ) -> Tuple["Layer", Mapping[Hashable, Set]]:
+        """Return a new Layer with only the tasks required to calculate `keys`.
+
+        In other words, remove unnecessary tasks from the layer.
+
+        Examples
+        --------
+        >>> d = Layer({'x': 1, 'y': (inc, 'x'), 'out': (add, 'x', 10)})  # doctest: +SKIP
+        >>> d.cull({'out'})  # doctest: +SKIP
+        {'x': 1, 'out': (add, 'x', 10)}
+
+        Returns
+        -------
+        layer: Layer
+            Culled layer
+        """
+        deps = self.get_dependencies(all_hlg_keys)
+
+        if len(keys) == len(self):
+            return self, deps  # Nothing to cull if preserving all existing keys
+
+        ret_deps = {}
+        seen = set()
+        out = {}
+        work = keys.copy()
+        while work:
+            k = work.pop()
+            out[k] = self[k]
+            ret_deps[k] = deps[k]
+            for d in deps[k]:
+                if d not in seen:
+                    if d in self:
+                        seen.add(d)
+                        work.add(d)
+
+        return BasicLayer(out), ret_deps
+
+    def get_dependencies(self, all_hlg_keys: Container) -> Mapping[Hashable, Set]:
+        """Get dependencies of all keys in the layer
+
+        Parameters
+        ----------
+        all_hlg_keys : Container
+            All keys in the high level graph.
+
+        Returns
+        -------
+        map: Mapping
+            A map that maps each key in the layer to its dependencies
+        """
+        return {k: keys_in_tasks(all_hlg_keys, [v]) for k, v in self.items()}
+
+
+class BasicLayer(Layer):
+    """Basic implementation of `Layer`
+
+    Parameters
+    ----------
+    mapping : Mapping
+        The mapping between keys and tasks, typically a dask graph.
+    dependencies : Mapping[Hashable, Set], optional
+        Mapping between keys and their dependencies
+    global_dependencies: Set, optional
+        Set of dependencies that all keys in the layer depend on. Notice,
+        the set might also contain literals that will be ignored.
```

Comment:
This no longer appears to be used.  Can we safely drop it?

Reply:
It is still used by `Blockwise` for the global (non-blockwise) keys: https://github.com/dask/dask/blob/2589faffec8c7b694ad6e7b1015accf51d8f1ba8/dask/blockwise.py#L229

Reply:
Ah I see, I was grepping for the keyword term and didn't see anything come up.  I should have looked for positional arg use.  Thanks for clarifying